Implement interrupts for 12864 LCD encoder with RAMPS 1.4 configuration

Joe321 opened the pull request:
A configuration item named USING_12864_WITH_ENCODER_INTERRUPTS
was added to the board_ramps.h include file to enable interrupt usage.

These changes require adding a jumper from D33 to D18, and one from
D31 to D19.  These can be done from the display connector to pins on
the RAMPS 1.4 board.  I cut a female jumper cable in half, soldered
the wire to the display adapter, and used the female end to attach to
the correct pin on the RAMPS 1.4 board.

D33 and D31 will still be set to inputs and receive the signal, but
they are not used.  Interrupts are enabled for D18 and D19, their
pull-ups are enabled, and the signal is read from the new pins.

Code was written to process the interrupts and handle menus.  Turning
the knob quicker will increment float and long values quicker.

The button press was not moved to an interrupt.  It still must be held
sufficiently long for the polling code to recognize the button press.

Commit:
Implement interrupts for 12864 LCD encoder with RAMPS 1.4 configuration

A configuration item named USING_12864_WITH_ENCODER_INTERRUPTS
was added to the board_ramps.h include file to enable interrupt usage.

These changes require adding a jumper from D33 to D18, and one from
D31 to D19.  These can be done from the display connector to pins on
the RAMPS 1.4 board.  I cut a female jumper cable in half, soldered
the wire to the display adapter, and used the female end to attach to
the correct pin on the RAMPS 1.4 board.

D33 and D31 will still be set to inputs and receive the signal, but
they are not used.  Interrupts are enabled for D18 and D19, their
pull-ups are enabled, and the signal is read from the new pins.

Code was written to process the interrupts and handle menus.  Turning
the knob quicker will increment float and long values quicker.

The button press was not moved to an interrupt.  It still must be held
sufficiently long for the polling code to recognize the button press.

i-make-robots commented:
What happens if interrupts fire while the motors are moving?

> On Jan 2, 2021, at 4:30 PM, Joe321 <notifications@github.com> wrote:
> 
> ﻿
> A configuration item named USING_12864_WITH_ENCODER_INTERRUPTS
> was added to the board_ramps.h include file to enable interrupt usage.
> 
> These changes require adding a jumper from D33 to D18, and one from
> D31 to D19. These can be done from the display connector to pins on
> the RAMPS 1.4 board. I cut a female jumper cable in half, soldered
> the wire to the display adapter, and used the female end to attach to
> the correct pin on the RAMPS 1.4 board.
> 
> D33 and D31 will still be set to inputs and receive the signal, but
> they are not used. Interrupts are enabled for D18 and D19, their
> pull-ups are enabled, and the signal is read from the new pins.
> 
> Code was written to process the interrupts and handle menus. Turning
> the knob quicker will increment float and long values quicker.
> 
> The button press was not moved to an interrupt. It still must be held
> sufficiently long for the polling code to recognize the button press.
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/MarginallyClever/Makelangelo-firmware/pull/102
> 
> Commit Summary
> 
> Implement interrupts for 12864 LCD encoder with RAMPS 1.4 configuration
> File Changes
> 
> M board_ramps.h (25)
> M configure.h (6)
> M lcd.cpp (195)
> M lcd.h (2)
> Patch Links:
> 
> https://github.com/MarginallyClever/Makelangelo-firmware/pull/102.patch
> https://github.com/MarginallyClever/Makelangelo-firmware/pull/102.diff
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

Joe321 commented:
I haven't dug that far into the code, and this is my first time using stepper motors in a project, so I could have misunderstood how things work here.  I thought the stepper motors just needed N pulses on the Step pin to move N steps and that the timing of the pulses didn't matter much.  If interrupts occurred, the pulses frequency would slow for a short duration (and step pin may be held high or lows slightly longer), but they would continue when the interrupt completes.  Is there precise timing required for driving that pin? 

The servo though, that needs good PWM?  That might be an issue depending on the servo code.

The encoder interrupts could potentially be disabled and re-enabled to avoid any time critical code sections.

i-make-robots commented on lcd.cpp:
```diff
@@ -430,12 +434,16 @@ void LCD_drive_menu() {
+  
   float offset[3];
-  offset[0] = axies[0].pos + lcd_turn > 0 ? 1 : -1;
+  offset[0] = axies[0].pos + (lcd_turns_to_process * multiplier);
```
lcd_turns_to_processs could be 10.  the if() and the multiplier wouldn't be needed every loop.

Joe321 replied:
lcd_turns_to_process depends on the number of turns made on the knob before the LCD_update routine runs.  There could be multiple interrupts because someone it turning the knob quickly, and the interrupt routine just counts the turns.  When LCD_update runs again (asynchronously to counting the turns), then LCD_update will store the current number of turns (lcd_turns) that occurred since the last time LCD_update ran in lcd_turns_to_process.  It doesn't just use the lcd_turn count since that could change during LCD_update processing if an interrupt occurs.  If the person is turning the knob quickly, then I made this code so that it would change the number quicker (10x).  They can do fine tuning by turning the knob slowly, or quick large changes (10x) by turning quickly.  This should just be intuitive for a user - they need a big change and turn quickly and the value just increments at 10x rate without them having to know  its happening.

i-make-robots commented on lcd.cpp:
```diff
@@ -485,14 +501,16 @@ void LCD_driveZ() {
+  if ( lcd_turns_to_process > 2 ) multiplier = 10;
+
+  // protect servo, don't drive beyond physical limits
+  float newF = feed_rate + (lcd_turns_to_process * multiplier);
```
when i turn positive, multiplier=10.  when i turn negative, multiplier=1.  this is intentional?

Joe321 replied:
nope, that needs to be fixed.  That should be an  abs compare so it does 10x for either direction.

i-make-robots commented on lcd.cpp:
```diff
@@ -802,12 +826,17 @@ void LCD_update() {
+
+      // reset.  must come after first (*current_menu)() because it might be used there, 
+      // and it should only be used once there if nextPage makes this loop.  (damn globals...)
+      lcd_turn -= lcd_turns_to_process; 
```
this is now inside a loop.  was outside.  might be run more times than intended.

Joe321 replied:
It needs to be inside the loop so the value isn't used multiple times inside of "current_menu".  That was a bug in the original code.  The lcd_turns_to_process value is set to 0, so that current_menu only uses a non-zero lcd_turns_to_process once.  And have to adjust lcd_turn before lcd_turns_to_process is zeroed out.

i-make-robots commented on lcd.cpp:
```diff
@@ -958,6 +1036,13 @@ void LCD_setup() {
+#ifdef USING_12864_WITH_ENCODER_INTERRUPTS
+  // set original pins as inputs, but do not enable their pull-ups
+  // new pins that are interruptable will enable pull-ups
+  pinMode(BTN_EN1_OLD, INPUT);
```
These are defined, connections are soldered (as per previous code comment) and then never used?

Joe321 replied:
They will remain connected because of the ramps 1.4 shield connection, so they need to be configured as inputs.  But they aren't used otherwise.

i-make-robots commented on board_ramps.h:
```diff
@@ -58,22 +58,35 @@
-#define BTN_EN1            31
-#define BTN_EN2            33
+#ifdef USING_12864_WITH_ENCODER_INTERRUPTS
+#define BTN_EN1            18    // solder a jumper from D33 on the display connector to D18 on the RAMPS 1.4 board
```
Why are we soldering here?  the _old goes low when the old en1 would go high, right?  I'd rather define an _ACTIVE as LOW or HIGH and instead of (digitalRead(BTN_EN1) == LOW) use (digitalRead(BTN_EN1) == BTN_ACTIVE)

Joe321 replied:
The ramps 1.4 "shield" board layout plugs into the D31 and D33 pins.  Those will still be connected, but don't have interrupts, so I have solder a jumper wires (no header pin I can get to) and then connect the jumpers to D18 and D19 that have interrupts.  The D31 and D33 pins will still be electrically connected, so they are still set to inputs, but the their pull-ups aren't enabled.

As for the == LOW logic... that was the way the existing code did the logic so I used it to be consistent.  The logic level didn't change from before, just using different pins with interrupts.

Joe321 replied:
![PXL_20210104_001033538](https://user-images.githubusercontent.com/7231870/103492326-184f2400-4df8-11eb-9019-62f244e34cbf.jpg)

Commit:
Fixes related to 128x64 display

- Add "abs" to check for quick knob turns per PR so that negative scrolls work quickly too
- Wrap M117 message when written to screen so the entire message can be read
- Fix screen location calculation for writes so the left most column is used
- Fix bug in LCD_setStatusMessage that could result in memory corruption

i-make-robots commented:
Hello!  I'd like to get this merged and closed.  Please resolve the latest conflicts or updated me on the status.  Thank you!

Joe321 commented:
Looks like the baseline has diverged significantly from the code I made my modifications on back in January, and would take some effort to merge my changes into the new code.  Wouldn't have time to take on that effort right now.
